File: frontend/src/features/auth/services/authService.ts
import { supabase, getEmailVerificationUrl, getPasswordResetUrl, getOtpRedirectUrl } from "@/shared/lib/supabase";
import apiClient from "@/shared/lib/axios";
import type { RegisterFormData, LoginFormData } from "../schemas/authSchemas";

export interface AuthUser {
  id: string;
  email: string;
  fullName: string;
}

export interface AuthResponse {
  accessToken: string;
  tokenType: string;
  user: AuthUser;
}

export interface MessageResponse {
  message: string;
}

export async function registerUser(data: RegisterFormData): Promise<AuthResponse> {
  // Use Supabase client directly for registration (handles session automatically)
  const { data: authData, error } = await supabase.auth.signUp({
    email: data.email,
    password: data.password,
    options: {
      data: { full_name: data.fullName },
      emailRedirectTo: getEmailVerificationUrl(),
    },
  });

  if (error) {
    console.error("[Auth] Registration failed:", {
      code: error.code,
      message: error.message,
      status: error.status,
      name: error.name,
    });
    const authError: AuthError = new Error(error.message || "Registration failed. Please try again.");
    authError.code = error.code;
    throw authError;
  }

  if (!authData.user) {
    throw new Error("Registration failed. Please try again.");
  }

  const user = authData.user;
  const session = authData.session;

  return {
    accessToken: session?.access_token ?? "",
    tokenType: "bearer",
    user: {
      id: user.id,
      email: user.email ?? data.email,
      fullName: user.user_metadata?.full_name ?? data.fullName,
    },
  };
}

export interface AuthError extends Error {
  code?: string;
}

export async function loginUser(data: LoginFormData): Promise<AuthResponse> {
  // Use Supabase client for auth (handles tokens automatically)
  const { data: authData, error } = await supabase.auth.signInWithPassword({
    email: data.email,
    password: data.password,
  });

  if (error) {
    const authError: AuthError = new Error(error.message || "Invalid email or password");
    authError.code = error.code;
    throw authError;
  }

  const session = authData.session;
  const user = authData.user;

  return {
    accessToken: session.access_token,
    tokenType: "bearer",
    user: {
      id: user.id,
      email: user.email ?? "",
      fullName: user.user_metadata?.full_name ?? "",
    },
  };
}

export async function logoutUser(): Promise<void> {
  const { error } = await supabase.auth.signOut();
  if (error) {
    throw new Error(error.message);
  }
}

export async function forgotPassword(email: string): Promise<MessageResponse> {
  // Call Supabase directly with the password reset redirect URL
  // This ensures the reset email contains the correct production redirect URL
  const { error } = await supabase.auth.resetPasswordForEmail(email, {
    redirectTo: getPasswordResetUrl(),
  });

  if (error) {
    console.error("[Auth] Password reset request failed:", {
      code: error.code,
      message: error.message,
      status: error.status,
    });
    // Return generic success message regardless of error to prevent email enumeration
    // (per requirement 1.4)
  }

  return { message: "If an account exists with that email, a password reset link has been sent." };
}

export async function resetPassword(
  token: string,
  newPassword: string
): Promise<MessageResponse> {
  const response = await apiClient.post<MessageResponse>("/auth/reset-password", {
    token,
    new_password: newPassword,
  });
  return response.data;
}

/**
 * Sign in with a magic link (OTP) sent to the user's email.
 * Uses the production-aware redirect URL for the magic link callback.
 */
export async function signInWithOtp(email: string): Promise<MessageResponse> {
  const { error } = await supabase.auth.signInWithOtp({
    email,
    options: {
      emailRedirectTo: getOtpRedirectUrl(),
    },
  });

  if (error) {
    console.error("[Auth] OTP sign-in failed:", {
      code: error.code,
      message: error.message,
      status: error.status,
    });
    const authError: AuthError = new Error(error.message || "Failed to send magic link.");
    authError.code = error.code;
    throw authError;
  }

  return { message: "A magic link has been sent to your email." };
}

File: backend/app/services/auth_service.py
"""Authentication service wrapping Supabase Auth methods."""

import logging

from fastapi import HTTPException, status

from app.integrations.supabase_client import get_supabase_client

logger = logging.getLogger(__name__)


class AuthService:
    """Service layer for authentication operations using Supabase Auth."""

    def __init__(self):
        self._client = get_supabase_client()

    def register(self, full_name: str, email: str, password: str) -> dict:
        """Register a new user account.

        Creates the user in Supabase Auth which also sends a verification email.

        Args:
            full_name: User's full name.
            email: User's email address.
            password: User's chosen password.

        Returns:
            Dictionary with access_token, token_type, and user info.

        Raises:
            HTTPException: 400 if registration fails (e.g., email already exists).
        """
        try:
            response = self._client.auth.sign_up(
                {
                    "email": email,
                    "password": password,
                    "options": {"data": {"full_name": full_name}},
                }
            )

            if not response.user:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Registration failed. Please try again.",
                )

            # Extract session info if available (auto-confirm enabled)
            access_token = ""
            if response.session:
                access_token = response.session.access_token

            return {
                "access_token": access_token,
                "token_type": "bearer",
                "user": {
                    "id": str(response.user.id),
                    "email": response.user.email or email,
                    "full_name": full_name,
                },
            }

        except HTTPException:
            raise
        except Exception as e:
            logger.error("Registration failed for email %s: %s", email, str(e))
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Registration failed. The email may already be registered.",
            ) from e

    def login(self, email: str, password: str) -> dict:
        """Authenticate a user with email and password.

        Args:
            email: User's email address.
            password: User's password.

        Returns:
            Dictionary with access_token, token_type, and user info.

        Raises:
            HTTPException: 401 with generic message if credentials are invalid.
        """
        try:
            response = self._client.auth.sign_in_with_password(
                {"email": email, "password": password}
            )

            if not response.user or not response.session:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Invalid email or password",
                )

            user_meta = response.user.user_metadata or {}

            return {
                "access_token": response.session.access_token,
                "token_type": "bearer",
                "user": {
                    "id": str(response.user.id),
                    "email": response.user.email or email,
                    "full_name": user_meta.get("full_name"),
                },
            }

        except HTTPException:
            raise
        except Exception as e:
            logger.warning("Login failed for email %s: %s", email, str(e))
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid email or password",
            ) from e

    def logout(self, access_token: str) -> None:
        """Invalidate the current user session.

        Args:
            access_token: The JWT access token to invalidate.

        Raises:
            HTTPException: 500 if logout fails unexpectedly.
        """
        try:
            self._client.auth.admin.sign_out(access_token)
        except Exception as e:
            logger.error("Logout failed: %s", str(e))
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Logout failed. Please try again.",
            ) from e

    def forgot_password(self, email: str) -> None:
        """Send a password reset email.

        Always succeeds regardless of whether the email is registered,
        to prevent email enumeration attacks.

        Args:
            email: Email address to send the reset link to.
        """
        try:
            from app.config import get_settings

            settings = get_settings()
            redirect_url = f"{settings.get_resolved_frontend_url()}/auth/callback"
            self._client.auth.reset_password_for_email(
                email, options={"redirect_to": redirect_url}
            )
        except Exception as e:
            # Silently catch all errors to prevent email enumeration
            logger.debug("Forgot password request for %s: %s", email, str(e))

    def reset_password(self, token: str, new_password: str) -> None:
        """Reset a user's password using a valid reset token.

        Args:
            token: The password reset token from the email link.
            new_password: The new password to set.

        Raises:
            HTTPException: 400 if the token is invalid or expired.
        """
        try:
            # Verify the token and get a session
            session_response = self._client.auth.verify_otp(
                {"token_hash": token, "type": "recovery"}
            )

            if not session_response.user:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Invalid or expired reset link",
                )

            # Update the user's password
            self._client.auth.admin.update_user_by_id(
                str(session_response.user.id),
                {"password": new_password},
            )

        except HTTPException:
            raise
        except Exception as e:
            logger.error("Password reset failed: %s", str(e))
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid or expired reset link",
            ) from e


def get_auth_service() -> AuthService:
    """Factory function for AuthService dependency injection."""
    return AuthService()
